packages/cli/src/WorkflowHelpers.ts

```diff
@@ -703,3 +704,56 @@ export function generateFailedExecutionFromError(
 		stoppedAt: new Date(),
 	};
 }
+
+export function validateWorkflowCredentialUsage(
```

Comment:
I rewrote this in order to understand it better. Adding it here just for reference. Untested.

```ts
/**
 * Get all nodes in a workflow where the node credential is not accessible to the user.
 */
function getNodesWithInaccessibleCreds(workflow: WorkflowEntity, userCredIds: string[]) {
	return workflow.nodes.filter((node) => {
		if (!node.credentials) return false;

		return Object.values(node.credentials)
			.map((nodeCred) => nodeCred.id)
			.some((nodeCredId) => nodeCredId !== null && !userCredIds.includes(nodeCredId));
	});
}

const isTamperingAttempt = (prevWorkflowNodeIds: string[], inaccessibleCredNodeId: string) =>
	!prevWorkflowNodeIds.includes(inaccessibleCredNodeId);

/**
 * Validate whether a new version of a workflow has been tampered with.
 */
export function validateNodeCredAccess(
	newWorkflow: WorkflowEntity,
	prevWorkflow: WorkflowEntity,
	userCreds: string[],
) {
	const nodesWithInaccesibleCreds = getNodesWithInaccessibleCreds(newWorkflow, userCreds);

	if (!nodesWithInaccesibleCreds.length) return newWorkflow;

	const prevWorkflowNodeIds = prevWorkflow.nodes.map((node) => node.id);

	nodesWithInaccesibleCreds.forEach((node, index) => {
		if (isTamperingAttempt(prevWorkflowNodeIds, node.id)) {
			throw new Error('Workflow contains new nodes with credentials inaccessible to user');
		}

		// as a precaution, replace node in new worfklow with node in existing workflow
		newWorkflow.nodes[index] = prevWorkflow.nodes.find((prevNode) => prevNode.id === node.id)!;
	});

	return newWorkflow;
}
```

Reply:
Thanks, I've improved the code based on your feedback and also fixed an issue I've detected.